smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/language/functions/endpoints/StandardRegionalEndpointsTrait.java

```diff
@@ -0,0 +1,167 @@
+    private final Map<String, List<RegionSpecialCase>> regionSpecialCases;
+
+    public StandardRegionalEndpointsTrait(StandardRegionalEndpointsTrait.Builder builder) {
+        super(ID, builder.getSourceLocation());
+        partitionSpecialCases = builder.partitionSpecialCases.copy();
+        regionSpecialCases = builder.regionSpecialCases.copy();
+    }
+
+    @Override
+    protected Node createNode() {
+        final Node partitionSpecialCasesNode = partitionSpecialCases.entrySet().stream()
+                .map(entry -> {
+                    List<Node> nodes = new ArrayList<>();
+                    for (PartitionSpecialCase partitionSpecialCase : entry.getValue()) {
+                        nodes.add(partitionSpecialCase.toNode());
+                    }
+                    return new AbstractMap.SimpleImmutableEntry<>(
+                            Node.from(entry.getKey()), Node.fromNodes(nodes));
+                })
+                .collect(ObjectNode.collect(Map.Entry::getKey, Map.Entry::getValue));
+
+        final Node regionSpecialCasesNode = regionSpecialCases.entrySet().stream()
+                .map(entry -> {
+                    List<Node> nodes = new ArrayList<>();
+                    for (RegionSpecialCase regionSpecialCase : entry.getValue()) {
+                        nodes.add(regionSpecialCase.toNode());
+                    }
+                    return new AbstractMap.SimpleImmutableEntry<>(
+                            Node.from(entry.getKey()), Node.fromNodes(nodes));
+                })
+                .collect(ObjectNode.collect(Map.Entry::getKey, Map.Entry::getValue));
+
+        return ObjectNode.objectNodeBuilder()
+                .sourceLocation(getSourceLocation())
+                .withMember(PARTITION_SPECIAL_CASES, partitionSpecialCasesNode)
+                .withMember(REGION_SPECIAL_CASES, regionSpecialCasesNode)
+                .build();
+    }
+
+    @Override
+    public SmithyBuilder<StandardRegionalEndpointsTrait> toBuilder() {
+        return new Builder()
+                .partitionSpecialCases(partitionSpecialCases)
+                .regionSpecialCases(regionSpecialCases);
+    }
+
+    public Map<String, List<PartitionSpecialCase>> getPartitionSpecialCases() {
+        return partitionSpecialCases;
+    }
+
+    public Map<String, List<RegionSpecialCase>> getRegionSpecialCases() {
+        return regionSpecialCases;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Provider extends AbstractTrait.Provider {
+        public Provider() {
+            super(ID);
+        }
+
+        @Override
+        public Trait createTrait(ShapeId target, Node value) {
+            ObjectNode objectNode = value.expectObjectNode();
+
+            StandardRegionalEndpointsTrait.Builder builder = builder()
+                    .sourceLocation(value);
+
+            objectNode
```

Comment:
This would be easier to read if it was done with a conditional to check for the presence of this member, then looping over its members and the nested array node.